rotkehlchen/accounting/cost_basis/base.py

```diff
@@ -136,44 +165,271 @@ def consume_result(self, used_amount: FVal) -> None:
+            taxable = True
+            if at_taxfree_period:
+                taxfree_amount += acquisition_event.remaining_amount
+                taxfree_bought_cost += acquisition_cost
+                taxable = False
+            else:
+                taxable_amount += acquisition_event.remaining_amount
+                taxable_bought_cost += acquisition_cost
 
-        self._acquisitions[0].remaining_amount -= used_amount
-        if self._acquisitions[0].remaining_amount == ZERO:
-            self._acquisitions.popleft()
+            log.debug(
+                'Spend uses up entire historical acquisition',
+                tax_status='TAX-FREE' if at_taxfree_period else 'TAXABLE',
+                bought_amount=acquisition_event.remaining_amount,
+                asset=spending_asset,
+                acquisition_rate=acquisition_event.rate,
+                profit_currency=settings.main_currency,
+                time=timestamp_to_date(acquisition_event.timestamp),
+            )
+            matched_acquisitions.append(MatchedAcquisition(
+                amount=acquisition_event.remaining_amount,
+                event=acquisition_event,
+                taxable=taxable,
+            ))
+            used_acquisitions.append(acquisition_event)
+            self.consume_result(acquisition_event.remaining_amount)
+            # and since this event is going to be removed, reduce its remaining to zero
+            acquisition_event.remaining_amount = ZERO
+
+        is_complete = True
+        if remaining_sold_amount != ZERO:
+            # if we still have sold amount but no acquisitions to satisfy it then we only
+            # found acquisitions to partially satisfy the sell
+            adjusted_amount = spending_amount - taxfree_amount
+            missing_acquisitions.append(
+                MissingAcquisition(
+                    asset=spending_asset,
+                    time=timestamp,
+                    found_amount=taxable_amount + taxfree_amount,
+                    missing_amount=remaining_sold_amount,
+                ),
+            )
+            taxable_amount = adjusted_amount
+            is_complete = False
+
+        return CostBasisInfo(
+            taxable_amount=taxable_amount,
+            taxable_bought_cost=taxable_bought_cost,
+            taxfree_bought_cost=taxfree_bought_cost,
+            matched_acquisitions=matched_acquisitions,
+            is_complete=is_complete,
+        )
 
     def __len__(self) -> int:
-        return len(self._acquisitions)
+        return len(self._acquisitions_heap)
+
 
+class FIFOCostBasisMethod(BaseCostBasisMethod):
+    """
+    Accounting in FIFO (first-in-first-out) method.
+    https://www.investopedia.com/terms/f/fifo.asp
+    """
+    def __init__(self) -> None:
+        super().__init__()
+        self._count = FVal(0)
```

Comment:
you forgot to use the constant here

```suggestion
        self._count = ZERO
```